goto symex member cleanup

karkhaz opened the pull request:
This patch-set aims to `const`-ify every member of `goto_symext` and related classes that only serve as a cache for an `optionst` option, and to make those members `protected` if appropriate. The fact that they were public and/or non-const was rather confusing.

The first commit in the patch-set allows to set options of an `optionst` while it is being constructed, so that other classes can read from the `optionst` object while *they* are being constructed. This means that we no longer need to do many of the `symex.member = options.get_option("member");` inside the bodies of a constructor, instead doing `member(options.get_option("member"))` inside the constructor for `goto_symext` and therefore making `member` const and protected.

karkhaz committed:
Set scratch_programt optionst on construction

This is so that the optionst object that gets passed to
scratch_programt::symex has the right options for initializing the symex
object.

karkhaz committed:
constify goto_symext::guard_identifier

This member is never written to.

karkhaz committed:
const & protect goto_symext::constant_propagation

This member is never written to and serves merely as a cache for a
front-end option, so this commit makes it const and protected.

karkhaz committed:
const and protect self_loops_to_assumptions

This goto_symext member is never written to and serves merely as a cache
for a front-end option, so this commit makes it const and protected.

karkhaz committed:
constify symex_bmct::record_coverage

This member is never written to and serves merely as a cache for a
front-end option, so this commit makes it const.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 0c909fd4c2c16c6d0ee26a310c471833f3f60b25).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88847152

smowton commented on src/goto-instrument/accelerate/scratch_program.h:
```diff
@@ -43,16 +43,17 @@ class scratch_programt:public goto_programt
       path_storage(),
-      options(),
+      options(
+        {{"propagation", true},
```
Why the new propagation setting?

smowton commented:
So symex currently has a weird mixture of directly querying the `optionst` it has stored by reference and querying it in the constructor and storing the results in fields (or previously, having its users do that for it). I suggest we should decide on one style and do that consistently -- personally I'd favour removing the `const optionst &options` from `goto_symext` entirely and querying and storing every option we're interested in during the constructor. That will lead to the more intuitive behaviour that the object is configured at construction time and subsequent changes to `options` will not affect it.

smowton commented:
Also I'm not a fan of the new constructor for `optionst` because it's awkward to specify e.g. a single string option (`optionst({}, {}, {}, {"hello", "world"})`). Instead how about constructing the `scratch_programt` options using a static function, so the initializer looks something like `options(get_default_options())`?

karkhaz committed:
Replace goto_symext::options with const members

goto_symext no longer has a reference to a optionst object. Instead, the
options that goto_symext needs to read from an optionst are read during
construction and cached in const members.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 56b4872c1b76513201aece41e95107a7e71601c9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89037402

karkhaz commented:
@smowton thanks for the comments, agreed with all of them. I've reverted the `optionst` constructor and replaced it with a commit to construct `scratch_programt`'s `optionst` with a static function. I've also added a new commit which introduces a bunch of members to `goto_symext`, caching the remaining options that it was using.

I notice that clang-format errors now cause the CI to fail. Is the preferred way to apply clang-format to have one single commit for the whole patch series as I've done above, or is it better to apply clang-format on each commit? `CODING_STANDARD.md` doesn't take a strong stance on the issue...

smowton commented on src/goto-instrument/accelerate/scratch_program.cpp:
```diff
@@ -203,3 +203,10 @@ void scratch_programt::append_loop(
   }
 }
+
+optionst scratch_programt::default_options;
```
Just declare this as a local in `get_default_options`

smowton reviewed (approved):
Change the static `default_options` into a local, otherwise LGTM, thanks for the improvements! re: clang-format it's best to do it per commit if possible, since that results in a more accurate `git blame`

karkhaz commented:
Thanks, I did as you asked and applied the clang-format changes to the right commits. (I've also squashed the clang-format changes into its commit in #3220).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 2fc5dabf8cdc19b138dc2a359c53f0b0aa31dc18).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89137416

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 65e017d15ed698e764d4df3d15507ca570cd8cd7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89630691
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 622940c06321fc7e00446475a7dec263adeaef36).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89632968
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (approved):
This looks good. Also, I have manually checked that this causes no trouble with test-gen. Compatibility warnings are due to some previous failure and can be disregarded.

chrisr-diffblue reviewed (approved):
That looks a lot cleaner to me - and thanks for constructing an easy-to-review PR :-)

karkhaz commented:
@chrisr-diffblue I'm trying to accumulate good karma to cash in for a 4k line PR that is on the horizon...:upside_down_face:
